### PR TITLE
NPC Persistence and Basic Entity Spawning (FIX)

### PR DESCRIPTION
**Pull request by kardall**
When copying over my files after a fork, apparently I forgot the WorldDatabase file (It only opened the worldserver project apparently)... SO...

Original: 
This is only the functionality of actually getting NPCs in the world through !entity add|del commands.

The NPCs are dead on spawn, which is the next step. I don't know if they will 'be resurrected' when the health is restored after re-initialization on server boot, so i wouldn't recommend going nuts spawning mobs yet.

Known Issue: The graphical display and the ID of the actual creature are not really ... right. The name is right but the models don't seem to match up for some reason on certain things.

### CODE REVIEW

File: Source/NexusForever.Database.World/WorldDatabase.cs [hunk 1]
```diff
@@ -35,6 +35,23 @@ public void Migrate()
             }
         }
 
+        public void SaveEntities(List<EntityModel> entities)
+        {
+            using var context = new WorldContext(config);
```
Comment (MaxtorCoder) [inline Source/NexusForever.Database.World/WorldDatabase.cs:40]: Preferably do:
```cs
using (var context = new WorldContext(config))
{
    // CODE
}

Comment (MaxtorCoder) [reply Source/NexusForever.Database.World/WorldDatabase.cs:40]: Ignore this XD

File: Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs [hunk 2]
```diff
@@ -1,19 +1,106 @@
...
+            
+            var entity = new NonPlayer();
+            entity.Initialise(npcModel, context.Invoker.Position, context.Invoker.Rotation);
+            context.GetTargetOrInvoker<Player>().Map.EnqueueAdd(entity, context.Invoker.Position);
+
+            // Add New Entity to the EntityCache
+            var cache = EntityCacheManager.Instance.GetEntityCache((ushort)context.Invoker.Map.Entry.Id);
+            cache.AddEntity(npcModel);
+            var activeNPCGrid = context.Invoker.Map.GetGrid(context.Invoker.Position);
+            var l = cache.GetEntities(activeNPCGrid.Coord.X, activeNPCGrid.Coord.Z).ToList();
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs]: Full variable name

File: Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs [hunk 3]
```diff
@@ -1,19 +1,106 @@
...
+            }
+
+            var npcModel = new EntityModel()
+            {
+                Id = DatabaseManager.Instance.WorldDatabase.GetNewEntityId(),
+                Creature = unitId,
+                Area = (ushort)context.GetTargetOrInvoker<Player>().Zone.Id,
+                DisplayInfo = creatureEntry.Id,
+                OutfitInfo = (ushort)creatureEntry.Creature2OutfitGroupId,
+                X = context.Invoker.Position.X,
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs:188]: Try to make use `context.GetTargetOrInvoker<Player>()`

File: Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs [hunk 4]
```diff
@@ -1,19 +1,106 @@
...
+            }
+            
+            context.Target.RemoveFromMap();
+            var model = DatabaseManager.Instance.WorldDatabase
+                .GetEntities((ushort)context.Invoker.Map.Entry.Id)
+                .Where(x => x.Id == context.Target.EntityId)
+                .FirstOrDefault();
+            EntityCacheManager.Instance.GetEntityCache((ushort)context.Invoker.Map.Entry.Id).DelEntity(model);
+            DatabaseManager.Instance.WorldDatabase.RemoveEntity(model);
+
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Command/Handler/EntityCommandCategory.cs:277]: Remove new lines

File: Source/NexusForever.WorldServer/Game/Map/BaseMap.cs [hunk 5]
```diff
@@ -204,7 +204,7 @@ private MapGrid GetGrid(uint gridX, uint gridZ)
             return grids[gridZ * MapDefines.WorldGridCount + gridX];
         }
 
-        private MapGrid GetGrid(Vector3 vector)
+        public MapGrid GetGrid(Vector3 vector)
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Map/BaseMap.cs:212]: Function documentation when made public

File: Source/NexusForever.WorldServer/Game/Map/EntityCache.cs [hunk 6]
```diff
@@ -34,12 +34,21 @@ public void AddEntity(EntityModel model)
             entities[coord].Add(model);
         }
 
+        public void DelEntity(EntityModel model)
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Map/EntityCache.cs]: `public void DeleteEntity(EntityModel model)`

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

File: Source/NexusForever.Database.World/WorldDatabase.cs (hunk 1 above)
Comment (Fabi) [reply Source/NexusForever.Database.World/WorldDatabase.cs:40]: > Ignore this XD

I like the old style more too but yea... :D

Comment (MaxtorCoder) [reply Source/NexusForever.Database.World/WorldDatabase.cs:40]: You mean the current style? Or the style I posted 😂

**Review by kardall — COMMENTED** [review]
Edited according to reviews.

**Comment by MaxtorCoder** [comment]
> Edited according to reviews.

Looks good.

**Review by MaxtorCoder — APPROVED** [review]

**Pull request closed without merging**